docs: Add documentation for six pylint errors

ollie-iterators opened the pull request:
<!--
Thank you for submitting a PR to pylint!

To ease the process of reviewing your PR, do make sure to complete the following boxes.

- [*] Write a good description on what the PR does.
- [ ] Create a news fragment with `towncrier create <IssueNumber>.<type>` which will be
  included in the changelog. `<type>` can be one of: breaking, user_action, feature,
  new_check, removed_check, extension, false_positive, false_negative, bugfix, other, internal.
  If necessary you can write details or offer examples on how the new change is supposed to work.
- [ ] If you used multiple emails or multiple names when contributing, add your mails
      and preferred name in ``script/.contributors_aliases.json``
-->

## Type of Changes

<!-- Leave the corresponding lines for the applicable type of change: -->

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
| ✓   | :sparkles: New feature |
| ✓   | :hammer: Refactoring   |
| ✓   | :scroll: Docs          |

## Description

This adds documentation for the following errors in pylint from pylint-errors:

- bad-mcs-classmethod-argument
- bad-mcs-method-argument
- inconsistent-mro
- docstring-first-line-empty
- differing-param-doc
- differing-type-doc


<!-- If this PR references an issue without fixing it: -->

Refs [#5953](https://github.com/PyCQA/pylint/issues/5953)

<!-- If this PR fixes an issue, use the following to automatically close when we merge: -->

Closes #XXXX

Pierre-Sassoulas commented on doc/data/messages/b/bad-mcs-method-argument/details.rst:
```diff
@@ -1 +1,7 @@
-You can help us make the doc better `by contributing <https://github.com/PyCQA/pylint/issues/5953>`_ !
+Used when a metaclass method has a first argument named differently than the
+value specified in valid-classmethod-first-arg option (default to `cls`),
+recommended to easily differentiate them from regular instance methods.
+
+Weird enough but usually naming of the first argument is `mcs` and pylint
+documentation says so but when someone specifies it then the lib throws
+C0203 error.
```
There's a description in the message directly in the checker, you should probably modify that and regenerate the doc :)

ollie-iterators replied:
Changed documentation to what I found in "doc/user_guide/checkers/features.rst"

Pierre-Sassoulas commented on doc/whatsnew/fragments/0203.docs:
```diff
@@ -0,0 +1 @@
+Added documentation for C0203 from pylint-errors
```
No need to document this with fragments, I added the skip news label

ollie-iterators replied:
Removed the fragments

Pierre-Sassoulas commented on doc/data/messages/d/differing-param-doc/bad.py:
```diff
@@ -0,0 +1,10 @@
+def foo(x, y):
```
Could you use real world example without foo / bar, please ? It's more lively and understandable for beginner with actual examples.

ollie-iterators replied:
Changed to a real word example

Pierre-Sassoulas commented on doc/data/messages/d/differing-param-doc/bad.py:
```diff
@@ -0,0 +1,14 @@
+def leapYear(year):
```
```suggestion
def leap_year(year):
```
Let's respect pep8 / python norms as much as we can :)

ollie-iterators replied:
Fixed

Pierre-Sassoulas commented on doc/data/messages/d/differing-param-doc/good.py:
```diff
@@ -1 +1,13 @@
-# This is a placeholder for correct code for this message.
+def leapYear(year):
```
```suggestion
def leap_year(year):
```

ollie-iterators replied:
Fixed

Pierre-Sassoulas commented on doc/data/messages/b/bad-mcs-method-argument/details.rst:
```diff
@@ -1 +1,4 @@
-You can help us make the doc better `by contributing <https://github.com/PyCQA/pylint/issues/5953>`_ !
+*Metaclass method %s should have %s as first argument*
+  Used when a metaclass method has a first argument named differently than the
+  value specified in valid-classmethod-first-arg option (default to "cls"),
+  recommended to easily differentiate them from regular instance methods.
```
Check the generated result, we're already using what's in the checker's message's description automatically. If there's nothing to add we can remove the details.rst altogether

ollie-iterators replied:
The generated result from pylint was "Metaclass class method __new__ should have 'mcs' as first argumentPylint(C0204:bad-mcs-classmethod-argument)"

Pierre-Sassoulas commented:
You can generate the doc locally with ``tox -e docs``

ollie-iterators commented:
> You can generate the doc locally with `tox -e docs`

I just did that

Pierre-Sassoulas commented on doc/data/messages/b/boolean-datetime/bad.py:
```diff
@@ -0,0 +1 @@
+
```
We'd need an example of bad code for this message

ollie-iterators replied:
The error doesn't happen unless you use Python version 3.5

Pierre-Sassoulas replied:
Use py-version=3.5 like us f-string-used- ... in version that does not support it.

ollie-iterators replied:
Ok, I stopped changing boolean-datetime

Pierre-Sassoulas commented on doc/data/messages/b/bad-mcs-method-argument/bad.py:
```diff
@@ -0,0 +1,3 @@
+class Meta(type):
+    def __new__(some):
+        pass
```
There need to be a message raised by the bad exampele somewhere, you can launch the check with ``pytest doc -k bad-mcs``

ollie-iterators replied:
I found a message for the code

ollie-iterators replied:
I made an issue happen

Pierre-Sassoulas commented on doc/data/messages/c/cyclic-import/bad.py:
```diff
@@ -0,0 +1,9 @@
+cat > a.py <<EOF
```
This is not valid python, it can't work. I think cyclic-import is not the easiest message to document as it implicate multiple files.

ollie-iterators replied:
I got the documentation from pylint-errors, which is where I was supposed to get it from for: https://github.com/PyCQA/pylint/issues/5953

Pierre-Sassoulas replied:
We added some more checks and normalization around it. Most of the message are already implemented so the remaining one can be a little hard to do (like cyclic-import or duplicate code). It's ok if you don't want /can't handle the hard one :)

Pierre-Sassoulas reviewed (changes requested):
The documentation check fail seems genuine, could you look into it please ?

Pierre-Sassoulas commented:
Alternatively there's some of the check that can be merged already, if you create a new MR with only those it could prevent duplicate like in #8184

ollie-iterators commented:
> Alternatively there's some of the check that can be merged already, if you create a new MR with only those it could prevent duplicate like in #8184

I removed the documentation for cyclic-import. I assume that is the only one with the issue after I removed boolean-datetime as well.

ollie-iterators commented:
What is wrong with the base dependencies?

Pierre-Sassoulas commented:
Relaunched, that's a temporary github issue

ollie-iterators commented:
> Relaunched, that's a temporary github issue

So should I be able to commit the changes?

Pierre-Sassoulas commented:
It seems there's a genuine issue in the current version, you can use ``pytest doc`` to verify locally.

ollie-iterators commented:
When I ran that, I got errors for things that I did not work with, like differing-param-doc

ollie-iterators commented:
This is what showed up:
FAILED doc/test_messages_documentation.py::test_code_examples[inconsistent-mro-bad] - AssertionError: assert Counter() == Counter({(9, ...ent-mro'): 1})
FAILED doc/test_messages_documentation.py::test_code_examples[too-many-public-methods-bad] - AssertionError: There should be at least one warning raised for...
FAILED doc/test_messages_documentation.py::test_code_examples[missing-format-string-key-bad] - AssertionError: There should be at least one warning raised f...
FAILED doc/test_messages_documentation.py::test_code_examples[docstring-first-line-empty-bad] - AssertionError: There should be at least one warning raised ...
FAILED doc/test_messages_documentation.py::test_code_examples[differing-param-doc-bad] - AssertionError: There should be at least one warning raised for 'ba...
FAILED doc/test_messages_documentation.py::test_code_examples[differing-type-doc-bad] - AssertionError: There should be at least one warning raised for 'bad...
FAILED doc/test_messages_documentation.py::test_code_examples[wrong-spelling-in-docstring-good] - SystemExit: 32
FAILED doc/test_messages_documentation.py::test_code_examples[wrong-spelling-in-docstring-bad] - SystemExit: 32
FAILED doc/test_messages_documentation.py::test_code_examples[wrong-spelling-in-comment-good] - SystemExit: 32
FAILED doc/test_messages_documentation.py::test_code_examples[wrong-spelling-in-comment-bad] - SystemExit: 32

Pierre-Sassoulas commented:
The real one are:
 FAILED doc/test_messages_documentation.py::test_code_examples[differing-param-doc-bad]
FAILED doc/test_messages_documentation.py::test_code_examples[docstring-first-line-empty-bad]
FAILED doc/test_messages_documentation.py::test_code_examples[differing-type-doc-bad]
FAILED doc/test_messages_documentation.py::test_code_examples[missing-format-string-key-bad]
FAILED doc/test_messages_documentation.py::test_code_examples[too-many-public-methods-bad]
FAILED doc/test_messages_documentation.py::test_code_examples[inconsistent-mro-bad]

For the spelling you'd need to install pyenchant for it to work, you can do that or ignore it.

ollie-iterators commented:
What is wrong with the errors? I have the wrong variable name in the differing-param-doc docstring.

Pierre-Sassoulas commented on doc/data/messages/d/differing-param-doc/pylintrc:
```diff
@@ -0,0 +1,2 @@
+[DESIGN]
+differing-param-doc:yes
```
```suggestion
differing-param-doc=yes
```

Pierre-Sassoulas commented on doc/data/messages/d/docstring-first-line-empty/pylintrc:
```diff
@@ -0,0 +1,2 @@
+[DESIGN]
+docstring-first-line-empty:yes
```
```suggestion
docstring-first-line-empty=yes
```

ollie-iterators replied:
So why is this still failing? It literally has the first line empty in the docstring

Pierre-Sassoulas commented on doc/data/messages/m/missing-format-string-key/pylintrc:
```diff
@@ -0,0 +1,2 @@
+[DESIGN]
+missing-format-string-key:yes
```
```suggestion
missing-format-string-key=yes
```

Pierre-Sassoulas reviewed (changes requested):
pylintrc format is ini, not sure ``:`` works

Pierre-Sassoulas reviewed (changes requested):
You also need to load the extension raising the message, for example ``load-plugins = pylint.extensions.docparams``

Pierre-Sassoulas commented:
> And where do I find that information on that? I didn't see any information on that on the moving errors from pylint-errors to pylint issue.

It's in the page you're generating https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/missing-any-param-doc.html (You should probably generate locally with tox -e docs and check the result in the generated html)

ollie-iterators commented:
Also, where is the information about how to specify when an error shows up in pylint in the documentation?

ollie-iterators commented:
> > And where do I find that information on that? I didn't see any information on that on the moving errors from pylint-errors to pylint issue.
> 
> It's in the page you're generating https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/missing-any-param-doc.html (You should probably generate locally with tox -e docs and check the result in the generated html)

I have added that information to some of the pylint errors that I have added documentation for, and nothing is happening.

Pierre-Sassoulas reviewed (commented):
Please do not pile new messages on top of the other. I'd rather have 12 merge requests and merge them one by one rather than this.

ollie-iterators commented:
> Please do not pile new messages on top of the other. I'd rather have 12 merge requests and merge them one by one rather than this.

I'll remove the extra onces once it is explained how to fix the following error documentations:

- bad-mcs-classmethod-argument
- bad-mcs-method-argument
- differing-param-doc
- differing-type-doc
- docstring-first-line-empty
- inconsistent-mro

ollie-iterators commented:
> Please do not pile new messages on top of the other. I'd rather have 12 merge requests and merge them one by one rather than this.

Maybe you should explain better how to make issues with the docstrings work with pylint.

Pierre-Sassoulas commented on doc/data/messages/b/bad-mcs-classmethod-argument/details.rst:
```diff
@@ -1 +1,8 @@
+(default to `cls`), recommended to easily differentiate them from regular
+instance methods.
+
+Weird enough but usually naming of the first argument is `mcs` and pylint
+documentation says so but when someone specifies it then the lib throws
+C0204 error.
```
I don't understand this comment. If something seems odd to you, then open an issue so we can discuss it. It makes no sense to add your personal dissent to the documentation.

Pierre-Sassoulas commented on doc/data/messages/b/bad-mcs-classmethod-argument/details.rst:
```diff
@@ -1 +1,8 @@
-You can help us make the doc better `by contributing <https://github.com/PyCQA/pylint/issues/5953>`_ !
+Used when a metaclass class method has a first argument named differently
+than the value specified in valid-metaclass-classmethod-first-arg option
+(default to `cls`), recommended to easily differentiate them from regular
+instance methods.
```
This should be in the code description of bad-mcs-classmethod in the checker that will automatically end up in the documentation.

Pierre-Sassoulas commented on doc/data/messages/b/bad-mcs-method-argument/details.rst:
```diff
@@ -1 +1,4 @@
-You can help us make the doc better `by contributing <https://github.com/PyCQA/pylint/issues/5953>`_ !
+*Metaclass method __new__ should have 'mcs' as first argument*
+  Used when a metaclass method has a first argument named differently than the
+  value specified in valid-classmethod-first-arg option (default to "cls"),
+  recommended to easily differentiate them from regular instance methods.
```
Same remark than for  bad-mcs-classmethod

Pierre-Sassoulas commented on doc/data/messages/b/broken-collections-callable/pylintrc:
```diff
@@ -1,3 +1,3 @@
-[main]
+[DESIGN]
```
I don't think this should be modified ?

Pierre-Sassoulas replied:
```suggestion
[main]
```

Pierre-Sassoulas commented on doc/data/messages/i/inconsistent-mro/details.rst:
```diff
@@ -1 +1 @@
-You can help us make the doc better `by contributing <https://github.com/PyCQA/pylint/issues/5953>`_ !
+Used when a class has an inconsistent method resolution order.
```
I suppose the message description already say this

Pierre-Sassoulas commented on doc/user_guide/configuration/all-options.rst:
```diff
@@ -171,7 +171,7 @@ Standard Checkers
 *Minimum Python version to use for version dependent checks. Will default to the version used to run pylint.*
 
-**Default:**  ``(3, 10)``
+**Default:**  ``(3, 11)``
```
```suggestion
**Default:**  ``(3, 10)``
```

Pierre-Sassoulas commented on doc/user_guide/configuration/all-options.rst:
```diff
@@ -264,7 +264,7 @@ Standard Checkers
    persistent = true
 
-   py-version = [3, 10]
+   py-version = [3, 11]
```
```suggestion
   py-version = [3, 10]
```

Pierre-Sassoulas commented on doc/user_guide/configuration/all-options.rst:
```diff
@@ -1267,7 +1267,7 @@ Standard Checkers
 --spelling-dict
 """""""""""""""
-*Spelling dictionary name. Available dictionaries: en (aspell), en_AU (aspell), en_CA (aspell), en_GB (aspell), en_US (aspell).*
+*Spelling dictionary name. Available dictionaries: none. To make it work, install the 'python-enchant' package.*
```
```suggestion
*Spelling dictionary name. Available dictionaries: en (aspell), en_AU (aspell), en_CA (aspell), en_GB (aspell), en_US (aspell).*
```

Pierre-Sassoulas reviewed (changes requested):
Thank you for lowering the number of message :)

ollie-iterators commented:
What is the issue with the docstring errors?

DanielNoord commented on doc/data/messages/d/differing-param-doc/pylintrc:
```diff
@@ -0,0 +1,2 @@
+[main]
+load-extension: pylint.extensions.docparams
```
This is the incorrect format. Please look at how other examples do this.

In this case this should work:
```suggestion
load-extension = pylint.extensions.docparams
```

ollie-iterators replied:
Still showing errors

Pierre-Sassoulas commented:
Please stop changing the title, please just focus on the 3 messages that were passing the tests and take the review remarks into account until we merge, then open another merge request for the remaining messages.

ollie-iterators commented:
Fine, if you are going to not provide any help.

ollie-iterators commented:
> Please stop changing the title, please just focus on the 3 messages that were passing the tests and take the review remarks into account until we merge, then open another merge request for the remaining messages.

No

DanielNoord commented:
> > Please stop changing the title, please just focus on the 3 messages that were passing the tests and take the review remarks into account until we merge, then open another merge request for the remaining messages.
> 
> If you are going to be no help.

Please refrain from such harsh comments. We are all volunteers. We appreciate every minute somebody spends on making `pylint` or its documentation better, but spamming us with commits and false claims that things are broken only costs us time without bringing the PR any further.

I have fixed all 6 messages for now, but I think a review is still necessary as some are not really in the explicit style of some of the other examples. Also please stop merging all of your commits into the branch. It makes no sense to have this branch have 140 commits and just makes reviewing a pain.

ollie-iterators commented:
> > > Please stop changing the title, please just focus on the 3 messages that were passing the tests and take the review remarks into account until we merge, then open another merge request for the remaining messages.
> > 
> > 
> > If you are going to be no help.
> 
> Please refrain from such harsh comments. We are all volunteers. We appreciate every minute somebody spends on making `pylint` or its documentation better, but spamming us with commits and false claims that things are broken only costs us time without bringing the PR any further.
> 
> I have fixed all 6 messages for now, but I think a review is still necessary as some are not really in the explicit style of some of the other examples. Also please stop merging all of your commits into the branch. It makes no sense to have this branch have 140 commits and just makes reviewing a pain.

Ok, fine

ollie-iterators commented:
Why is missing-parentheses-for-call-in-test suddenly calling an error?

DanielNoord commented:
@ollie-iterators I have no idea what you are referring to and I'm going to close this PR. You blatantly ignored my previous comment about not keeping all commits around and this PR is getting out of hand.

I created https://github.com/PyCQA/pylint/pull/8202 to not let this work go to waste and all commits credit you as they should. You are welcome to create another PR with more examples if you wish but please make sure tests pass locally before opening the PR so not to consume to much of the limited time of maintainers.
